Prevent sum overflow in broadcast_object_list

suphoff opened the pull request:
Summary: broadcast_object_list casted the sum of all object lengths to int from long causing overflows.

Test Plan:
sandcastle
buck test mode/dev-nosan //caffe2/test/distributed:distributed_gloo_spawn

Differential Revision: D33281300



cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

rohan-varma reviewed (commented):
Looks great, although would be good to also have a unittest with the change.

Commit:
Prevent sum overflow in broadcast_object_list (#70336)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/70336

broadcast_object_list casted the sum of all object lengths to int from long causing overflows.

Test Plan:
Increased size of Tensor used in object transfers to have  >2GB storage requirement (in distributed_test.py)

Without fix the length will overflow and the program will request a negative sized Tensor:
```
RuntimeError: Trying to create tensor with negative dimension -2147482417: [-2147482417]
```
With fix it will pass the test.

Test used on server with GPUs:

buck test  mode/dev-nosan //caffe2/test/distributed:distributed_nccl_spawn --local -- broadcast_object

Differential Revision: D33281300

fbshipit-source-id: 704a240e362b9e5f06f80109171d0b69c7c2f124

suo commented:
this PR broke master https://www.torch-ci.com/commit/807f9a828cb841b5688cf0011e49994ac58a87c1 I am reverting
